### ISSUE TITLE
"out of range" error for some SRFIs

### ISSUE THREAD

**Issue by siiky** [ursetto/chicken-doc#6]
`chicken-doc` fails with an "out of range" error when asking for some SRFIs' docs. These are the offending SRFIs:

SRFI-105
SRFI-130
SRFI-135
SRFI-178
SRFI-189
SRFI-196

```
$ chicken-doc srfi-189
path: (srfi-189)

Error: (u8vector-ref) out of range
#u8(255 255 255 255 255 255 255 255 255 255 255 255 255 255 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 120 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 254 255 255 255 255 255 31 0 0 0 0 0 0 0 0 0 63 0 0 0 0 0 0 248 255 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 192 255 255 63 0 0 0 0 2 0 0 0 255 255 255 7 0 0 0 0 0 0 0 0 0 0 192 255 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...
```

It seems `u8vector`s are not used directly in `chicken-doc`, so the problem may be in a dependency, but I have no other clue about that...

---

While looking for other SRFIs I noticed that SRFI-111 also fails, but with a different error (probably unrelated):

```
$ chicken-doc srfi-111
path: (srfi-111)

Error: (caar) bad argument type: ()
```

**Comment by ursetto** [comment]
I'll take a look. This can happen when importing via `chicken-doc-admin`, if the source is badly formed, but shouldn't happen on display.

**Comment by ursetto** [comment]
I suspect this to be an issue external to chicken-doc-text, caused by the fmt egg and interaction with Unicode characters like U+2618 (☞). Some chars work fine (srfi-45 é) but others do not. The text engine relies heavily on wrapping, so perhaps the character width table is at fault.

Specifically, for chars between #x0300 and #x3029, it uses the low-non-spacing-chars table to determine char width. The table (a u8vector) exactly matches the u8vector-ref error.

**Comment by siiky** [comment]
With your pointers I went digging a bit, and I don't understand much of what's going on, but I found that indeed `low-non-spacing-chars` doesn't have enough elements (958):

```scm
; fmt-unicode-chicken.scm:95
((<= #x0300 ci #x3029)
 (let* ((i (- ci #x0300))
        (byte (quotient i 8))
        (off (remainder i 8)))
   (if (zero? (bitwise-and (u8vector-ref low-non-spacing-chars byte)
                           (arithmetic-shift 1 off)))
       1
       0)))
```

```
#;1> (quotient (- #x3029 #x0300) 8)
1445
#;2> (- (quotient (- #x3029 #x0300) 8) 958)
487
```

**Comment by ursetto** [comment]
Yeah, the table is only valid for 0x300-0x20ef, really it needs to be capped at that point. I'm trying to regenerate the table now and fix a couple other bugs.

**Comment by ursetto** [comment]
I have fixed the issue in fmt-unicode but need to talk to upstream before applying.
